processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java

```diff
@@ -306,27 +305,22 @@ public String apply(@Nullable AggregatorFactory input)
       @Override
       public byte[] computeCacheKey(TopNQuery query)
       {
-        final byte[] dimensionSpecBytes = query.getDimensionSpec().getCacheKey();
-        final byte[] metricSpecBytes = query.getTopNMetricSpec().getCacheKey();
-
-        final DimFilter dimFilter = query.getDimensionsFilter();
-        final byte[] filterBytes = dimFilter == null ? new byte[]{} : dimFilter.getCacheKey();
-        final byte[] aggregatorBytes = QueryCacheHelper.computeAggregatorBytes(query.getAggregatorSpecs());
-        final byte[] granularityBytes = query.getGranularity().cacheKey();
-
-        return ByteBuffer
-            .allocate(
-                1 + dimensionSpecBytes.length + metricSpecBytes.length + 4 +
-                granularityBytes.length + filterBytes.length + aggregatorBytes.length
-            )
-            .put(TOPN_QUERY)
-            .put(dimensionSpecBytes)
-            .put(metricSpecBytes)
-            .put(Ints.toByteArray(query.getThreshold()))
-            .put(granularityBytes)
-            .put(filterBytes)
-            .put(aggregatorBytes)
-            .array();
+        final CacheKeyBuilder builder = new CacheKeyBuilder(TOPN_QUERY)
+            .appendCacheable(query.getDimensionSpec())
+            .appendCacheable(query.getTopNMetricSpec())
+            .appendInt(query.getThreshold())
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimensionsFilter())
+            .appendCacheablesIgnoringOrder(query.getAggregatorSpecs());
+
+        final List<PostAggregator> postAggregators = prunePostAggregators(query);
```

Comment:
oh, duh, because the post-aggs can be calculated from the cached results.

Are they? My mind is drawing a blank there. Do post-aggs get calculated AFTER cached results?